cedar-lean/Cedar/Spec/Ext/Datetime.lean

```diff
@@ -0,0 +1,119 @@
+/-
+ Copyright Cedar Contributors
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-/
+
+import Cedar.Data.Int64
+
+/-! This file defines Cedar datetime values and functions. -/
+
+namespace Cedar.Spec.Ext
+
+open Cedar.Data
+
+/--
+  A duration value is measured in milliseconds and constructed from a duration string.
+  A duration string is a concatenated sequence of quantity-unit pairs where the quantity
+  is a positive integer and unit is one of the following:
+    - `d` (for days)
+    - `h` (for hours)
+    - `m` (for minutes)
+    - `s` (for seconds)
+    - `ms` (for milliseconds)
+
+  Duration strings are required to be ordered from largest unit to smallest
+  unit, and contain one quantity per unit. Units with zero quantity may be
+  omitted.
+
+  A duration may be negative. Negative duration strings must begin with `-`.
+-/
+abbrev Duration := Int64
+
+namespace Datetime
+
+def MILLISECONDS_PER_SECOND: Int := 1000
+def MILLISECONDS_PER_MINUTE: Int := 60000
+def MILLISECONDS_PER_HOUR: Int := 360000
+def MILLISECONDS_PER_DAY: Int := 86400000
+
+----- Definitions -----
+
+def duration? (i : Int) : Option Duration :=
+  Int64.mk? i
+
+def durationUnits? (i: Int) (suffix: String) : Option Duration :=
+  match suffix with
+  | "ms" => duration? i
+  | "s" => duration? (i * MILLISECONDS_PER_SECOND)
+  | "m" => duration? (i * MILLISECONDS_PER_MINUTE)
+  | "h" => duration? (i * MILLISECONDS_PER_HOUR)
+  | "d" => duration? (i * MILLISECONDS_PER_DAY)
+  | _ => none
+
+def unitsToMilliseconds (days hours minutes second milliseconds: Int) : Int :=
+  days * MILLISECONDS_PER_DAY +
+  hours * MILLISECONDS_PER_HOUR +
+  minutes * MILLISECONDS_PER_MINUTE +
+  second * MILLISECONDS_PER_SECOND +
+  milliseconds
+
+def isNegativeDuration (str: String) : Bool × String :=
+  match str.front with
+  | '-' => (true, str.drop 1)
+  | _   => (false, str)
+
+def addOptionDurations (a b : Option Duration) : Option Duration :=
+  match a, b with
+  | some durationA, some durationB => some (Int64.add? durationA durationB).get!
+  | some duration, none => some duration
+  | none, some duration => some duration
+  | none, none => none
+
+def parseUnit? (str : String) (suffix: String) : Option Duration × String :=
```

Comment:
This code is not distinguishing between the case where the suffix is absent and when there is a error, which makes it hard to correctly handle overflow behaviors in the rest of the implementation.

I suggest changing this function to have the following signature:

```
def parseUnit? (str : String) (suffix: String) : Option Duration
```

The behavior of the function would then be as follows:
- If the suffix is absent, return `some 0`.
- If the suffix is present, return`durationUnits? v suffix` where `v` is the result of successfully parsing a Nat value via `String.toNat?`.  Note that we don't want to use `String.toInt!` for parsing, because that will miss errors.  And because we're handling `-` separately, at this point in the code, we want to parse numbers as Nats rather than Ints.
- Otherwise, return `none`.

This means that `parseUnsignedDuration?` can use the `do` notation to bail out of parse errors immediately, and it can just add up the results of all unit returns if all the `parseUnit?` calls are successful.

Reply:
Thanks for the suggestion! I've rewritten the code to follow this behavior, please let me know if there's a discrepancy anywhere.